Refactor tests.testDCClient as dynamicconfig.MemoryClient

prathyushpv opened the pull request:
## What changed?
Refactor testDCClient from tests package as dynamicconfig.MemoryClient.

## Why?
This DC Client is useful for testing in other packages.

## How did you test it?
Existing unit tests

## Potential risks
None

## Documentation

## Is hotfix candidate?
No

dnr reviewed (commented):
What do you think about this instead:
- Move this to `common/dynamicconfig` and call it `MemoryClient`
- Get rid of the "fallback" feature that's not used (just return nil instead)
- Make a `SetValue[ByKey]` on `MemoryClient` not use `testing.T` (so that `MemoryClient` doesn't depend on `testing`), but then:
- Make `OverrideValue[ByKey]` on either `temporalImpl` or `FunctionalTestBase` that does hook into testing/suite to do cleanup

prathyushpv commented:
> What do you think about this instead:
> 
> * Move this to `common/dynamicconfig` and call it `MemoryClient`
> * Get rid of the "fallback" feature that's not used (just return nil instead)
> * Make a `SetValue[ByKey]` on `MemoryClient` not use `testing.T` (so that `MemoryClient` doesn't depend on `testing`), but then:
> * Make `OverrideValue[ByKey]` on either `temporalImpl` or `FunctionalTestBase` that does hook into testing/suite to do cleanup

Yeah that makes more sense.
I have refactored the code to do this. I moved cleanup logic to temporalImpl. But I kept fallback. This could be useful if a test wants to read configs from a file and then override a few keys.

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -0,0 +1,77 @@
+func (d *MemoryClient) GetRawValue(name Key) (any, bool) {
+	d.RLock()
+	defer d.RUnlock()
+	v, ok := d.overrides[name]
+	return v, ok
```
```suggestion
	return d.overrides[name]
```
should work?

prathyushpv replied:
TIL it is not possible to return d.overrides[name]. This will only return the value and not (value, ok).

dnr replied:
doh, you're right

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+type MemoryClient struct {
+	sync.RWMutex
```
I like to do
```suggestion
	lock sync.RWMutex
```
so that clients don't have direct access to the lock

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func (d *MemoryClient) OverrideValue(setting GenericSetting, value any) {
+	d.overrides[setting.Key()] = value
```
missing lock?  maybe call OverrideValueByKey instead?

prathyushpv replied:
Done

dnr commented on tests/advanced_visibility.go:
```diff
@@ -2022,9 +2022,8 @@ func (s *AdvancedVisibilitySuite) Test_BuildIdIndexedOnCompletion_UnversionedWor
-	dc := s.testCluster.host.dcClient
-	dc.OverrideValue(s.T(), dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
-	dc.OverrideValue(s.T(), dynamicconfig.MatchingNumTaskqueueWritePartitions, 1)
+	s.testCluster.host.OverrideDCValue(s.T(), dynamicconfig.MatchingNumTaskqueueReadPartitions, 1)
```
I want to simplify this more so you don't need the call to T() and the long path, but that can be in another PR

prathyushpv replied:
Do you propose adding OverrideDCValue() to FunctionalTestBase?

prathyushpv replied:
I have made that change as well. Let me know if that is what you were thinking about.

dnr replied:
yup, exactly what I was thinking, except you can move the s.T() also

prathyushpv replied:
Done

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -0,0 +1,77 @@
+
+func (d *MemoryClient) GetValue(name Key) []ConstrainedValue {
+	if val, ok := d.GetRawValue(name); ok {
+		return []ConstrainedValue{{Value: val}}
```
I know it's not used right now, but we could do this to make it more flexible:
if `val` already has type `[]ConstrainedValue` then return it directly, otherwise wrap it like this.

prathyushpv replied:
Done

prathyushpv replied:
Now that we allow either []ConstrainedValue or a Value in the map, I am confused about how GetRawValue() should behave. Tests expect it to return a value. But we might have []ConstrainedValue. Is it okay to return the first value in the list?

dnr replied:
commented on the new code.. let's get rid of GetRawValue

prathyushpv replied:
Done

dnr commented on tests/onebox.go:
```diff
@@ -221,11 +247,17 @@ func newTemporal(t *testing.T, params *TemporalParams) *temporalImpl {
 	}
-	impl.overrideHistoryDynamicConfig(t, testDCClient)
+	for k, v := range staticOverrides {
+		impl.overrideDynamicConfigByKey(t, k, v)
```
I'm not sure if this is better or worse, but just to consider:
The `fallback` for the memory client can be `dynamicconfig.StaticClient(staticOverrides)`, then you only need to set the ones from params and history ones.

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// NewMemoryDCClient - returns a memory based dynamic config client
+func NewMemoryDCClient(fallback Client) *MemoryClient {
```
```suggestion
func NewMemoryClient(fallback Client) *MemoryClient {
```
"DC" is in the package name, don't need to repeat it

prathyushpv replied:
Done

dnr commented on tests/onebox.go:
```diff
@@ -998,3 +1030,20 @@ func sdkClientFactoryProvider(
+
+func (c *temporalImpl) overrideDynamicConfigByKey(t *testing.T, name dynamicconfig.Key, value any) {
+	constrainedValues := c.dcClient.GetValue(name)
+	existed := len(constrainedValues) == 0
```
shouldn't this be != 0?

prathyushpv replied:
Yes!

prathyushpv replied:
Done

dnr commented on tests/onebox.go:
```diff
@@ -998,3 +1030,20 @@ func sdkClientFactoryProvider(
+			for _, v := range constrainedValues {
+				values = append(values, v)
+			}
+			c.dcClient.OverrideValueByKey(name, values)
```
doesn't this not work without the type check in GetValue? it'll re-wrap the list...
but also, why copy it to a `[]any` at all?

prathyushpv replied:
Yeah, that was a mistake.

prathyushpv replied:
Done

prathyushpv commented on common/dynamicconfig/memory_client.go:
```diff
@@ -0,0 +1,85 @@
+	defer d.lock.RUnlock()
+	v, ok := d.overrides[name]
+	if value, ok := v.([]ConstrainedValue); ok {
+		return value[0].Value, true
```
Now that we allow either []ConstrainedValue or a Value in the map, I am confused about how GetRawValue() should behave. Tests expect it to return a value. But we might have []ConstrainedValue. Is it okay to return the first value in the list?

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -29,32 +29,40 @@ import (
+	defer d.lock.RUnlock()
 	v, ok := d.overrides[name]
+	if value, ok := v.([]ConstrainedValue); ok {
+		return value[0].Value, true
```
this seems wrong, you can't just pick the first one.. also this whole function is unused now?

oh, I see where this is used. uugh, that's messy. I think you can just use GetValue there and get rid of the "must override" check. pick the first value from the ConstrainedValues with len(Constraints)==0, and panic if there is none.

(the idea was that those tests always should be explicit about partition count since the behavior can be pretty different with 1 vs many partitions, and I wanted most tests to use 1 partition. but I think it doesn't really matter that much.)

prathyushpv replied:
Done

dnr commented on common/dynamicconfig/memory_client.go:
```diff
@@ -29,32 +29,40 @@ import (
+		}
+		return []ConstrainedValue{{Value: v}}
 	}
 	return d.fallback.GetValue(name)
```
Should probably not hold the lock while calling the fallback? I guess it's unlikely to matter in practice.

But I'm also still questioning the need for the fallback. Specifically: the next step is to make MemoryClient implement NotifyingClient so subscriptions work instantaneously. Doing that with fallback is a lot more complicated since you have to check if that one can notify, subscribe to it, pass through callbacks, etc. And having a fallback but having subscriptions only work on MemoryClient but not the fallback is confusing semantics. So I might still lean towards dropping it, if there's not a very clear use case.

prathyushpv replied:
Right. Now I think it is better to remove fallback.

dnr commented on tests/functional_test_base.go:
```diff
@@ -437,3 +437,7 @@ func (s *FunctionalTestBase) registerArchivalNamespace(archivalNamespace string)
 	return err
 }
+
+func (s *FunctionalTestBase) overrideDynamicConfig(t *testing.T, setting dynamicconfig.GenericSetting, value any) {
```
great, this is a lot cleaner. I think it can be improved even more to have this call s.T() instead of having the caller do it.

prathyushpv replied:
Done

dnr commented on tests/nexus_workflow_test.go:
```diff
@@ -45,11 +45,11 @@ import (
 	"go.temporal.io/api/workflowservice/v1"
 	"go.temporal.io/sdk/client"
 	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/server/common/dynamicconfig"
```
move to server block

prathyushpv replied:
Done

dnr commented on tests/nexus_workflow_test.go:
```diff
@@ -772,7 +772,7 @@ func (s *ClientFunctionalSuite) TestNexusOperationAsyncCompletionErrors() {
 	s.NoError(err)
 
 	s.T().Run("ConfigDisabled", func(t *testing.T) {
```
use `s.Run()` instead of `s.T().Run()`, then `s.T()` will return the subtest `t` in the test, and you can safely call `s.T()` from `overrideDynamicConfig`

prathyushpv replied:
Oh that was the issue! Thanks! I tried removing t argument from OverrideDCValue() and reverted that because of this test failing.

dnr commented on tests/onebox.go:
```diff
@@ -998,3 +1030,15 @@ func sdkClientFactoryProvider(
+	c.dcClient.OverrideValueByKey(name, value)
+	t.Cleanup(func() {
+		if len(existingValues) > 0 {
+			c.dcClient.OverrideValueByKey(name, existingValues)
```
just thinking.. if there was a static override with a plain value, this takes the return value from GetValue, and passes it back to OverrideValueByKey, it'll get wrapped in a []ConstrainedValue, so not identical to how it was before. that's totally fine though, it'll work as expected (as long as Cleanup is called like a stack, which it is).

prathyushpv replied:
Yeah, it should be fine. We could wrap the value in []ConstrainedValue when it is first inserted so that it stays consistent.

dnr commented on tests/namespace_delete.go:
```diff
@@ -112,11 +112,10 @@ func (s *namespaceTestSuite) SetupTest() {
-	dc.OverrideValue(s.T(), dynamicconfig.ValidateUTF8FailRPCRequest, false)
-	dc.OverrideValue(s.T(), dynamicconfig.ValidateUTF8FailRPCResponse, false)
-	dc.OverrideValue(s.T(), dynamicconfig.ValidateUTF8FailPersistence, false)
+	s.cluster.host.OverrideDCValue(s.T(), dynamicconfig.ValidateUTF8FailRPCRequest, false)
+	s.cluster.host.OverrideDCValue(s.T(), dynamicconfig.ValidateUTF8FailRPCResponse, false)
+	s.cluster.host.OverrideDCValue(s.T(), dynamicconfig.ValidateUTF8FailPersistence, false)
```
should these use `s.overrideDynamicConfig` also?

prathyushpv replied:
This test suite does not use FunctionalTestBase. Probably it can be refactored to use it.

dnr commented on tests/functional_test_base.go:
```diff
@@ -437,3 +437,7 @@ func (s *FunctionalTestBase) registerArchivalNamespace(archivalNamespace string)
 	return err
 }
+
+func (s *FunctionalTestBase) overrideDynamicConfig(setting dynamicconfig.GenericSetting, value any) {
```
I'm wondering if this should be public so it can be used from other packages, e.g. the ndc/xdc tests. Currently some xdc tests do override dynamic config, but they manage multiple clusters so they call temporalImpl.OverrideDCValue directly, which is fine.

prathyushpv replied:
No tests in other packages were embedding FunctionalTestBase so far. But you are right, it should be possible.

dnr reviewed (approved):
Thanks, this is great!

Just a couple more comments but feel free to fix (or not) and merge
